include/nbl/builtin/hlsl/emulated/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,622 @@
+            }
+        }
+
+        emulated_float64_t operator*(float rhs)
+        {
+            return _static_cast<this_t>(data) * create(rhs);
+        }
+
+        emulated_float64_t operator/(const emulated_float64_t rhs) NBL_CONST_MEMBER_FUNC
+        {
+            if (FlushDenormToZero)
+            {
+                if (!FastMath && (tgmath::isnan<uint64_t>(data) || tgmath::isnan<uint64_t>(rhs.data)))
+                    return bit_cast<this_t>(ieee754::traits<float64_t>::quietNaN);
+
+                const uint64_t sign = (data ^ rhs.data) & ieee754::traits<float64_t>::signMask;
+
+                if (!FastMath && impl::isZero(rhs.data))
+                    return bit_cast<this_t>(ieee754::traits<float64_t>::quietNaN | sign);
+                if (!FastMath && impl::areBothInfinity(data, rhs.data))
+                    return bit_cast<this_t>(ieee754::traits<float64_t>::quietNaN | ieee754::traits<float64_t>::signMask);
+                if (!FastMath && tgmath::isinf(data))
+                    return bit_cast<this_t>(ieee754::traits<float64_t>::inf | sign);
+                if (!FastMath && tgmath::isinf(rhs.data))
+                    return bit_cast<this_t>(0ull | sign);
+                if (!FastMath && impl::isZero(rhs.data))
+                    return bit_cast<this_t>(ieee754::traits<float64_t>::quietNaN | sign);
+
+                int lhsBiasedExp = ieee754::extractBiasedExponent(data);
+                int rhsBiasedExp = ieee754::extractBiasedExponent(rhs.data);
+
+                uint64_t lhsData = impl::flushDenormToZero(lhsBiasedExp, data);
+                uint64_t rhsData = impl::flushDenormToZero(rhsBiasedExp, rhs.data);
+
+                const uint64_t lhsRealMantissa = (ieee754::extractMantissa(lhsData) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt));
+                const uint64_t rhsRealMantissa = ieee754::extractMantissa(rhsData) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt);
+
+                int exp = lhsBiasedExp - rhsBiasedExp + int(ieee754::traits<float64_t>::exponentBias);
+
+                uint64_t2 lhsMantissaShifted = impl::shiftMantissaLeftBy53(lhsRealMantissa);
+                uint64_t mantissa = impl::divmod128by64(lhsMantissaShifted.x, lhsMantissaShifted.y, rhsRealMantissa);
+
+                while (mantissa < (1ull << 52))
+                {
+                    mantissa <<= 1;
+                    exp--;
+                }
+
+                mantissa &= ieee754::traits<float64_t>::mantissaMask;
+
+                return bit_cast<this_t>(impl::assembleFloat64(sign, uint64_t(exp) << ieee754::traits<float64_t>::mantissaBitCnt, mantissa));
+            }
+            else
+            {
+                //static_assert(false, "not implemented yet");
+                return bit_cast<this_t>(0xdeadbeefbadcaffeull);
+            }
+        }
+
+        // relational operators
+        // TODO: should `FlushDenormToZero` affect relational operators?
```

Comment:
No, why:
- relational operators are `const` in c++ and will be in HLSL as soon as it supports const-member functions, so you'd have an operator making a 0 valued tmp out of both operands but not saving result causing confusion
- if any function that PRODUCES an `emulated_float64_t` (such as operators, and `create`) cares about `FlushDenormToZero` and implements it, then by the time we're in the `const` methods any operand will be already flushed to 0 (you could even runtime assert it!)